### ISSUE TITLE
Changing filtering parameters when building COB

### ISSUE THREAD

**Issue by MeeshCompBio** [schae234/Camoco#23]
Hello Rob,

When building a COB with the co.COB.py from_table script:

    def from_table(cls, filename, name, description,
                   refgen, rawtype=None, sep='\t', **kwargs):
        '''
            Build a COB Object from an FPKM or Micrarray CSV. This is a
            convenience method which handles reading in of tables.
            Files need to have gene names as the first column and
            accession (i.e. experiment) names as the first row. All
            kwargs will be passed to COB.from_DataFrame(...). See
            docstring there for option descriptions.
            Parameters
            ----------
            filename : str (path)
                the path to the FPKM table in csv or tsv
            name : str
                Name of the dataset stored in camoco database
            description : str
                Short string describing the dataset
            refgen : camoco.RefGen
                A Camoco refgen object which describes the reference
                genome referred to by the genes in the dataset. This
                is cross references during import so we can pull information
                about genes we are interested in during analysis.
            rawtype : str (default: None)
                This is noted here to reinforce the impotance of the rawtype
                passed to camoco.Expr.from_DataFrame. See docs there for
                more information.
            sep : str (default: \\t)
                Specifies the delimiter of the file referenced by the
                filename parameter.
            **kwargs : key value pairs
                additional parameters passed to subsequent methods.
            Returns
            -------
                a COB object
        '''
        return cls.from_DataFrame(
            pd.read_table(
                filename,sep=sep,
                compression='infer'
            ),
            name,description,refgen,
            rawtype=rawtype,**kwargs
        )

There does not seem to be a way to modify the filtering criteria such as min FPKM, or at least one gene in a tissue having an FPKM of 5 or greater. I did find a script in the co.EXPR.py script that is used on quality control for expression values which is my input for the from_table function and I am wondering if modifying that function will modify the filtering criteria for creating a COB from the from_table function?

        def _quality_control(self, min_expr=1, max_gene_missing_data=0.2, \
        min_single_sample_expr=5, max_accession_missing_data=0.3, \
        membership=None, dry_run=False, **kwargs):
        '''
            Perform Quality Control on raw expression data. This method filters
            genes based on membership to some RefGen instance, filters based on
            a minimum FPKM or equivalent expression value, filters out genes
            and accessions with too much missing data, filters out genes which
            are lowly expressed (do not have at least one accession that meets
            an FPKM threshold, i.e. likely presence absense). See parameters
            for more details.
            Parameters
            ----------
            min_expr : int (default: 1)
                FPKM (or equivalent) values under this threshold will be set to
                NaN and not used during correlation calculations.
            max_gene_missing_data : float (default: 0.2)
                Maximum percentage missing data a gene can have. Genes under
                this are removed from dataset.
            min_single_sample_expr : int (default: 5)
                Genes that do not have a single accession having an expression
                value above this threshold are removed from analysis. These are
                likely presence/absence and will not have a strong coexpression
                pattern.
            max_accession_missing_data : float (default: 0.5)
                maximum percentage missing data an accession (experiment) can
                have before it is removed.
            membership : RefGen
                Genes which are not contained within this RefGen will be
                removed. Note: this could also be another object that will
                implement an interface that will check to see if gene ids are
                contained within it i.e. a set of gene ids.
            dry_run : bool (default: False)
                Used in testing to speed up calculations. Limits the QC
                dataframe to only have 5000 genes.
        '''
        self.log('------------Quality Control')
        df = self.expr()
        # remember how we set the flags
        self._global('qc_min_expr', min_expr)
        self._global('qc_max_gene_missing_data', max_gene_missing_data)
        self._global('qc_min_single_sample_expr', min_single_sample_expr)
        self._global('qc_max_accession_missing_data', max_accession_missing_data)
        # Retrieve raw data as a data frame
        self.log('Raw Starting set: {} genes {} accessions'.format(
            len(df.index), len(df.columns))
        )
        # Remember why we remove certain genes
        # If TRUE it passes, if FALSE it fails!!!
        qc_gene = pd.DataFrame({'has_id':True}, index=df.index)
        qc_accession = pd.DataFrame({'has_id':True}, index=df.columns)

        # -----------------------------------------
        # Gene Membership test
        if not membership:
            membership = self.refgen
        self._global('qc_membership', str(membership))
        self.log("Filtering out genes not in {}", membership)
        qc_gene['pass_membership'] = [x in membership for x in df.index]

        # -----------------------------------------
        # Set minimum FPKM threshold
        self.log("Filtering out expression values lower than {}", min_expr)
        df_flt = df.copy()
        df_flt[df < min_expr] = np.nan
        df = df_flt
        # -----------------------------------------
        # Gene Missing Data Test
        self.log(
            "Filtering out genes with > {} missing data",
            max_gene_missing_data
        )
        qc_gene['pass_missing_data'] = df.apply(
            lambda x : ((sum(np.isnan(x))) < len(x)*max_gene_missing_data),
            axis=1
        )
        # -----------------------------------------
        # Gene Min Expression Test
        # filter out genes which do not meet a minimum expr
        # threshold in at least one sample
        self.log(("Filtering out genes which "
                  "do not have one sample above {}"), min_single_sample_expr)
        qc_gene['pass_min_expression'] = df.apply(
            lambda x: any(x >= min_single_sample_expr),
            axis=1 # 1 is column
        )
        qc_gene['PASS_ALL'] = qc_gene.apply(
            lambda row: np.all(row), axis=1
        )
        df = df.loc[qc_gene['PASS_ALL'], :] 
        # -----------------------------------------
        # Filter out ACCESSIONS with too much missing data
        self.log("Filtering out accessions with > {} missing data", max_accession_missing_data)

        qc_accession['pass_missing_data'] = df.apply(
            lambda col : (
                ((sum(np.isnan(col)) / len(col)) <= max_accession_missing_data)
            ),
            axis=0 # 0 is columns
        )
        # Update the total QC passing column
        qc_accession['PASS_ALL'] = qc_accession.apply(
            lambda row: np.all(row), axis=1
        )
        df = df.loc[:, qc_accession['PASS_ALL']] 
        # Update the database
        self.hdf5['qc_accession'] = qc_accession
        self.hdf5['qc_gene'] = qc_gene
        # Report your findings
        self.log('Gene Removal:\n{}', str(qc_gene.apply(sum, axis=0)))
        self.log('Accession Removal:\n{}', str(qc_accession.apply(sum, axis=0)))
        # Also report a breakdown by chromosome
        qc_gene = qc_gene[qc_gene['pass_membership']]
        qc_gene['chrom'] = [self.refgen[x].chrom for x in qc_gene.index]
        self.log('Genes passing QC by chromosome:\n{}',
            str(qc_gene.groupby('chrom').aggregate(sum, axis=0))
        )
        # update the df to reflect only genes/accession passing QC
        self.log('Kept: {} genes {} accessions'.format(len(df.index), len(df.columns)))
        if dry_run:
            # If dry run, take first 100 rows of QC
            self.log.warn("Dry Run")
            df = df.iloc[0:5000, :]
        self._update_values(df, 'quality_control')

Thank you

**Comment by schae234** [comment]
Hey Meesh,

The code is a little convoluted, but its possible to pass arguments from the `from_table` function all the way down to the `_quality_control` function where you found the parameter you want to change. 

As you pointed out, since the COB object inherits all of the methods from the Expr object, it also has an `_quality_control` method which works internally on the `_expr` data structure. 

In the `from_table` method, you see all the regular parameters for reading in the FPKM table, but on [lines 909 and 910](https://github.com/schae234/Camoco/blob/806ee0077f9c3176726a187202bd578469057775/camoco/COB.py#L909-L910) there is a special argument called a key-word argument which is just an additional list of key-value pairs (e.g. a=1, b=2, etc) which get passed around and used when they are needed. Since `from_table` is just a wrapper to `from_DataFrame` you see that it just reads in the file and calls the other method, but here is where the magic happens, it tacks on that `**kwags` at the [end ](https://github.com/schae234/Camoco/blob/806ee0077f9c3176726a187202bd578469057775/camoco/COB.py#L916-L923). So whatever extra parameters you put in the function call gets passed onto the next function call, as long as they keep passing the special `**kwarg` object on (the `**kwarg` notation is a python thing for which there is tons of docs for).

So, `from_table` calls `from_DataFrame` which [calls the Expr version](https://github.com/schae234/Camoco/blob/806ee0077f9c3176726a187202bd578469057775/camoco/COB.py#L916-L923) of `from_DataFrame` using the super() method, and inside of the [Expr method for from_DataFrame](https://github.com/schae234/Camoco/blob/806ee0077f9c3176726a187202bd578469057775/camoco/Expr.py#L651), there is a [call to `_quality_control`](https://github.com/schae234/Camoco/blob/806ee0077f9c3176726a187202bd578469057775/camoco/Expr.py#L711) and they all pass on this `**kwarg` object which means that if at the very top, when you call the cob method for `from_table` and pass in a value for `_quality_control` such as `min_expr=0.00000001`, it'll eventually get put in the right place.

You can see this in action in some of the [test code which builds COB objects](https://github.com/schae234/Camoco/blob/806ee0077f9c3176726a187202bd578469057775/tests/conftest.py#L80-L96)

So, long story short, you should be able to specify the parameters for the quality control function in the from_* methods. The code has a little bit of misdirection, but when its all set up, it makes it nice to be able to pass around these keyword arguments.

Does that make sense?
Rob

**Comment by schae234** [comment]
Sorry, i accidentally hit close.

**Comment by MeeshCompBio** [comment]
Thanks that does make sense. What if I wanted to remove a range of values such as everything between 0.0000000001 and 2?

**Comment by schae234** [comment]
You would need to add a function. This should be relatively straight forward to implement, the only problem right now is that the QC stages happen no matte what, so we would need to be able to choose which QC metric to run (the one implemented right now versus filtering between 0.000...0001 and 2). 

Maybe it would be easier to add back the original FPKMs which were 0? We should sit down and talk more about it, but if you wanted to poke around, the code you would need to start with is the code inside of the Expr [`from_DataFrame`](https://github.com/schae234/Camoco/blob/806ee0077f9c3176726a187202bd578469057775/camoco/Expr.py#L709-L720) method. There are a couple of steps there which iteratively perform QC and normalization,  you could add another step that recovers zeros or does your own QC method. Check out those methods, each of them are set up in a similar way. They each get a dataframe with `self.expr()` method, which gets the current *good* dataset, do some sort of modification and then call [`self._update_values`](https://github.com/schae234/Camoco/blob/806ee0077f9c3176726a187202bd578469057775/camoco/Expr.py#L198) which takes in a dataframe and a short log file message and updates the current *good* dataset inside the database. So if you were to go grab the expr dataframe and make a call to `_update_values` then update the coexpression network. That code is in the [`_from_Expr`](https://github.com/schae234/Camoco/blob/806ee0077f9c3176726a187202bd578469057775/camoco/COB.py#L814-L837) method, which has a terrible name which should probably be changed. :grimacing:

**Comment by schae234** [comment]
@MeeshCompBio I'm going to close this issue. I think we should try pair programming again and work through this problem. We can switch keyboards too :+1:
